### ISSUE TITLE
Marlin M>1 (default-on) stales the capture-safe m==1 invariant and reds 3 fused-SwiGLU tests via non-byte-identical split-K

### ISSUE THREAD

**Issue by justinchuby** [justinchuby/onnx-genai#1405]
## Summary

Default-on Marlin (`ONNX_GENAI_MARLIN_M_GT_1`, default **on** via `marlin_m_gt_1_enabled()` in `crates/onnx-runtime-ep-cuda/src/kernels/marlin_gemm.rs:128`) has two coupled problems on the fused gate/up `MatMulNBits` path:

1. **It stales the capture-safe invariant.** For `M>1`, `run_f16_gate_up_swiglu` takes the Marlin path and advertises `last_call_capture_safe = warm` (true after warmup), while the long-standing assertion in `matmul_nbits.rs` says **only `m == 1` may be advertised capture-safe**. So current default behaviour already contradicts that invariant.

2. **Its split-K reduction is documented non-byte-identical (and possibly non-deterministic),** which reds fused-SwiGLU parity tests on **pristine `main`** that are byte-exact whenever Marlin M>1 is disabled.

**Not fixing Marlin here** — filing so the owner of the Marlin path can decide. This is the invariant-staling + non-byte-identity finding, named precisely and left in place.

## Hardware / invocation

- RTX 4060 Laptop 8 GB, driver 591.55, CUDA 13.1 runtime (WDDM). CUDA runtime on PATH:
  ```powershell
  $nv = "$env:LOCALAPPDATA\anaconda3\Lib\site-packages\nvidia"
  $env:PATH = "$nv\cu13\bin\x86_64;$nv\cudnn\bin;$env:PATH"
  ```
- Full lib suite: `cargo test -p onnx-runtime-ep-cuda --features cuda --lib -- --test-threads=1`
- Toggle: prefix with `$env:ONNX_GENAI_MARLIN_M_GT_1='0'` to disable Marlin M>1.

## A/B evidence (pristine `main` @ f72700a8, RTX 4060, full lib suite)

| | Marlin default-**ON** | Marlin M>1 **OFF** (`=0`) |
|---|---|---|
| pristine `main` | **404 passed / 6 failed / 21 ignored** | **407 passed / 3 failed / 21 ignored** |

The three tests that flip **red → green** when Marlin M>1 is disabled — i.e. the Marlin split-K artifacts:

1. `kernels::matmul_nbits::tests::fp16_gate_up_swiglu_is_bit_exact_to_two_op_path`
   ```
   paired gate/up SwiGLU diverged at M=5, K=896, N=4864, row=0, column=951:
   fused=1132.0 reference=1133.0   (1 fp16 ULP)
   ```
   With Marlin M>1 off, the fused M=5 path is the tiled prefill GEMM, byte-exact to the two-op reference.

2. `kernels::matmul_nbits::tests::fused_gate_up_swiglu_rmsnorm_is_bit_exact_to_two_step_path`
   ```
   assertion `left == right` failed: only M=1 decode may be advertised capture-safe
    right: false
   ```
   Marlin's warm path advertises `capture_safe = true` at M>1, tripping the `m == 1` invariant.

3. `kernels::matmul_nbits::tests::fused_gate_up_swiglu_rmsnorm_zero_points_is_bit_exact_to_two_step_path` — same `only M=1 decode may be advertised capture-safe`.

The remaining 3 failures under Marlin-off (`graph::host_excursion...`, `graph::mid_segment_capture_failure...`, `group_query_attention::reference_scores_path_gates_on_dtype_and_decode_support`) are **not** Marlin-related — the two `graph::` tests are flaky and the GQA one is a separate pre-existing failure (persists under both configs).

## Why this matters

- The `m == 1` capture-safe invariant is treated as load-bearing by other work, but Marlin already violates it by default — any code that trusts the invariant is reasoning against stale behaviour.
- Marlin split-K being non-byte-identical/non-deterministic means M>1 gate/up decode logits are not reproducible under the default config. PR #1404 replaces the **rmsnorm** M>1 decode with a deterministic per-row GEMV loop that is 0 ULP from the two-op reference at M==1 — a reproducibility improvement — but it does **not** touch Marlin itself.

## Ask

Owner of the Marlin path to decide: either make the split-K path byte-identical/deterministic, or reconcile the capture-safe advertisement with the documented invariant (e.g. advertise capture-safe only where the result is deterministic). Corrects the premise of #1305 (these are Marlin artifacts, not a fundamental RTX 4060 rounding floor).

**Comment by justinchuby** [comment]
Part 2 of this issue — the byte-exactness tests going red under default-on Marlin M>1 — was addressed in #1509 (squashed as `6bc9bfb2`).

The conclusion there matched this issue's diagnosis: direct Marlin M>1 runs an `mma.sync` tensor-core reduction whose accumulation order differs from the CUDA-core tiled GEMM, so bit-exactness between the two is not an achievable target and the test was asserting something the dispatcher cannot deliver. Rather than relaxing the tolerance globally, the assertions are now gated on the dispatcher's own selection predicates, so each shape is held to the bound appropriate to the kernel that actually ran:

- shapes that elect Marlin prefill: <= 2 ULP
- split-K RMS at `M == 1`: <= 3 ULP
- everything that stays on the same reduction tree as the reference: still asserted **bit-identical**

`fp16_gate_up_swiglu_is_bit_exact_to_two_op_path` and `fp16_gate_up_swiglu_rmsnorm_two_op_ulp_bound_sweep` are green on an A100 with default settings after that change (full `--lib` suite: 426 passed / 0 failed).

**Part 1 is untouched and this issue should stay open for it.** The `M>1` Marlin path still advertises `last_call_capture_safe = warm` while the in-tree invariant says only `m == 1` may be advertised capture-safe. Nothing in #1509 reconciles those; it only stopped the numerics tests from asserting an impossible equality. The invariant question is still for the Marlin path owner to decide.

Note also that this GPU is an A100 (sm80) rather than the RTX 4060 in the report, so the ULP figures above are not a direct A/B against the numbers in the issue body.
